### PR TITLE
Memoise module paths

### PR DESCRIPTION
**Pull request by jonludlam**
The aim here is to reduce the amount of memory used when compiling long chains of aliases, which currently causes problems.

### CODE REVIEW

File: src/xref2/compile.ml [hunk 1]
```diff
@@ -16,9 +16,9 @@ let type_path : Env.t -> Paths.Path.Type.t -> Paths.Path.Type.t =
   match p with
   | `Resolved _ -> p
   | _ -> (
-      let cp = Component.Of_Lang.(type_path empty p) in
+      let cp = Component.Of_Lang.(type_path (empty ()) p) in
       match Tools.resolve_type_path env cp with
-      | Ok p' -> `Resolved (Cpath.resolved_type_path_of_cpath p')
+      | Ok p' -> `Resolved Lang_of.(Path.resolved_type (empty ()) p')
```
Comment (Julow) [inline src/xref2/compile.ml:21]: Why this change ?

**Review by Julow — COMMENTED** [review]
Could it be a global memo ? There's a lot of `empty ()` so potentially a lot of memos that aren't shared.

**Comment by jonludlam** [comment]
I considered a global memo, but it's all quite contextual - so converting from the same `Lang.* -> Component.*` will often give different results - whereas the reverse, `Component.* -> Lang.*` will likely cause a lot of leakage from memoised results that are subsequently irrelevant .

File: src/xref2/compile.ml (hunk 1 above)
Comment (jonludlam) [reply src/xref2/compile.ml:21]: We had two ways to convert from `Cpath.*` to `Path.*` - I deleted the ones in `Cpath` and added the memoisation to the other (in `Lang_of`).

**Comment by jonludlam** [comment]
In case it wasn't very obvious from the diff, this can make a startling difference (in the pathological case the test case is testing) - the size of the `odocl` file reduced from 36 megs to 24k :-) (https://github.com/ocaml/odoc/pull/799/commits/c7b327b014caaad2fc2df90e6ea4a25532f92e3e#diff-ad34e4af41233ad07cbd0172dc46cc763e6e50a0b0d403bf371ca7b7768a4e8dL5)

**Review by Julow — APPROVED** [review]
Awesome !
